[SETTING/#8] Security 및 OpenFeign 설정

yummygyudon opened the pull request:
<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closed #8 

## Work Description ✏️
- Spring Clouf OpenFeign 기본 설정
- Spring Security 기본 설정

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

### JwtProvider 추상화
> 4b6daba47217a04cd2a8ed86ac24b72ab3680c72

우선 인증 Jwt 정책에 대한 구현은 @hyunw9 님이 담당하기로 하여 인터페이스로 추상화하여 행위만 정의한 후,
`AuthenticationJwtProvider` 객체를 선언한 해놓은 상태입니다!!

**[Provider를 추상화하여 분리한 이유]**
- Jwt는 "인증"외 다른 관심사에서도 사용할 가능성이 높다고 생각을 했습니다. 이에 행위까지만 정의 & Generic을 통해 Provider의 관심사를 명시할 수 있도록 했습니다.
- 모든 Provider는 같은 행위를 수행할 수 있도록 강제할 수 있습니다.

<br/>

### OpenFeign Retry 제외
> c0b787f06400a7b662b288ec7116aeb0d4457ade

구현까지 완료하였으나 (조금 아깝긴 하네요...ㅠㅜ)
현재 인증 서버 분리 프로젝트에서 분명히 필요한 상황이 아니기 때문에 오버 엔지니어링이라고 판단하여 제외했습니다.

추후에 어플리케이션 안정성 측면에서 Retry 정책이 필요하게 된다면 추가하도록 하겠습니다.

<br/>

### 설정값(`.yml` / `.properties`) 관리 객체 
> 129f93e24776fb63bd372cbba8a890f888097f60

설정파일의 값을 객체로 바인딩하여 사용하는 방법은 상당히 다양합니다.
이 중 java 17 이상의 버전을 쓰는만큼 최대한 제공하는 API를 활용하고자 `Record` 타입으로 객체를 사용하고자 했습니다.

**[Record & `@ConfigurationProperties`를 사용한 이유]**
- 값이 매우 많을 경우, 모든 필드마다 `@Value` 어노테이션을 함께 등록해주는 비효율성 개선
- 불변성 보장 (Record 타입)
  - 기존에 불변성 보장을 위해선 일반 class에 `@ConstructorBinding` & `@RequiredArgsConstructor`이 필요했지만 이제 필요 없음
- inner Record를 통해 그룹핑 용이 (`@NestedConfigurationProperty` 사용 안해도 가능)
- `@ConfigurationProperties`이 지원하는 RelaxedBinding이 적용되어 보다 유연하게 값 바인딩 가능

**※ 단, 설정파일의 Key값과 매핑되기 때문에 field의 이름은 key이름과 통일해야 합니다**
**※ 단, Spring Application 클래스가 위치한 모듈에서도 해당 모듈만의 설정 파일이 있기 때문에 `@EnableConfigurationProperties` 명시가 필요합니다.**

yummygyudon committed:
[CHORE] Spring Cloud Open Feign Dependency 추가

yummygyudon committed:
[CHORE] Feign Test 환경 세팅

- `FeignAutoConfiguration.class`
  - Feign Context, FeignJacksonConfiguration 등 Feign의 기본적인 설정값이 존재하는 Configuration 집합을 활용합니다.
- `HttpMessageConvertersAutoConfiguration.class`
  - Feign에서 사용되어야 하는 Coverter Configuration을 활용합니다.

yummygyudon committed:
[CHORE] Application 모듈 등록

- 각 모듈에 구현되어 있는 구현체들 및 Configuration들을 사용하기 위해 각 모듈의 Root 인터페이스를 BasePackageClasses로 등록해주었습니다.

yummygyudon committed:
[CHORE] `core-api` 모듈 내 OpenFeign Dependency 추가

- Spring Application 내에서 Bean으로 등록되어 사용할 수 있어야 합니다.
- Spring Application Class에 `@EnableFeignClients`를 지정해주어야 합니다.
  - FeignClient는 기본적으로 Interface이기 때문에 Bean으로도 등록이 불가능합니다.

yummygyudon committed:
[FEAT] Open Feign 전역 Config 정의

- Log Level을 Local과 Test 환경에서만 Feign 통신의 Full 로그를 출력할 수 있도록 하여 디버깅을 용이하게 했습니다.
- Log가 상당히 길게 출력되기 때문에 Prod와 Dev 환경에서는 로그가 생성되지 않도록 NONE으로 설정했습니다.

yummygyudon committed:
[FEAT] Playground 전용 Feign Client 및 Config 정의

- Bearer 방식의 인증된 요청만 보낼 수 있도록 커스텀 Request Interceptor 구현하여 사용하도록 했습니다.
  - 타 클라이언트에서도 재사용 및 유연한 맞춤 구현을 할 수 abstract class로 구현했습니다.

yummygyudon committed:
[FEAT] Code Enum 통일성이 보장된 확장을 위해 Base Interface 구현

yummygyudon committed:
[FEAT] Base Exception 정의 & AuthException/TokenException 구현

- 반드시 생성자로 `BaseCode`를 받도록 강제하여 Type 보장을 시켰습니다.
  - 본 Exception을 상송받아 세분화할 경우, 각 Exception에서도 해당 관심사에 맞는 BaseCode implements Failure을 주입받도록 강제하는 방식으로 활용합니다.
 - 로깅 또한 해당 BaseException에 설정할 경우, 일괄적으로 적용할 수 있습니다.

yummygyudon committed:
[FEAT] 일관된 API Response 형식 보장을 위한 BaseResponse 구현

- 메서드명을 통해 생성 객체의 특징을 명확하게 하고자 했습니다.
  - 단순 `of`로만 정의하여 오버로딩할 경우, 사용자 측면에서 혼란이 생길 가능성을 고려했습니다.

yummygyudon committed:
[FEAT] `infra` 모듈 `client` 관심사 & `support` 모듈 `system` 관심사에 대한 상수 클래스 정의

- `final class` 정의를 통해 Util 용 객체임을 코드적으로 명시했습니다.
- 생성자 접근제어자를 private으로 설정함으로서 객체 생성을 방지했습니다.

yummygyudon committed:
[CHORE] 각 모듈에서의 Property 클래스 활성화

- `@EnableConfigurationProperties`를 통해 설정파일(.yml, .properties)의 값을 주입받은 Value 객체에 대해 등록하여 정상적으로 Spring Application에 등록되어 사용될 수 있도록 했습니다.

yummygyudon committed:
[FEAT] 확장 가능한 JwtProvider Interface 구현

- JWT의 경우, 인증 외 다른 관심사에서도 재사용할 가능성이 높은 도메인이기 때문에 확장성을 위해 제네릭을 활용한 인터페이스로 구현했습니다.
- token 정책 구현은 @hyunw9이 담당하기로 해서 우선 형식만 구현해놓았습니다.

yummygyudon committed:
[FIX] : BaseException 생성자 변경 및 하위 Exception 반영 & 관심사 세분화에 따른 패키지 변경

- Response에 대한 객체들이 많아짐에 따라 단순 `util` 패키지 안에 위치하는 것은 적절치 않다고 판단했습니다.
  - Request와 Response 모두 http 통신 관심사에 해당하기에 `http` 패키지로 세분화했습니다.
  - BaseResponse Record의 경우, `dto` 패키지에서 관리하도록 했습니다.
  - `http` 패키지에는 Generator & Util 객체를 위치시켰습니다.

yummygyudon committed:
[FEAT] JWT 인증 필터 & 인증 실패 필터 구현

- HttpServletRequest/Response 사용을 위해서 `spring-boot-web` Dependency 주입이 필요했습니다.

yummygyudon committed:
[FEAT] Spring Security Config 설정

- 환경별로 유연하게 세팅할 수 있도록 설정별로 메서드로 분리하여 조합했습니다.
  - Local/Test 환경에서는 `anyRequests().permitAll()` 설정으로 접속은 가능합니다.
  - JWT Auth Filter을 제거할 수 없기 때문에 Authorization Header 파싱 로직에서의 NPE 예외는 피할 수 없습니다.
  - [참고](https://stackoverflow.com/questions/46068433/spring-security-with-filters-permitall-not-working)

hyunw9 commented on infra/src/main/java/sopt/makers/authentication/infra/client/openfeign/config/BearerAuthRequestInterceptor.java:
```diff
@@ -0,0 +1,17 @@
+import feign.RequestTemplate;
+import lombok.Getter;
+
+@Getter
+public abstract class BearerAuthRequestInterceptor implements RequestInterceptor {
+
+    private final String token;
+
+    public BearerAuthRequestInterceptor(String token) {
+        this.token = token;
+    }
+
+    public abstract void apply(RequestTemplate template);
```
해당 interceptor 와 apply의 용도가 뭔지 여쭤봐도 될까요 ? 

feign으로 수신한 토큰에 대한 파싱을 인터셉트 하는 용도일까요 ? ? ?

yummygyudon replied:
RequestInterceptor 라는 이름처럼
Feign -> 외부 서버로 요청할 때, 요청 전에 Intercept해 본 Config 설정파일을 활용하는 Client의 요청엔 공통적으로 주입되도록 한거에요!!

Playground Token이 헤더에 항상 주입되어야 하니 위와 같이 Interceptor를 활용하여
자칫 Client 인터페이스 내 Feign 메서드마다 반복될 수 있는 헤더 주입 코드 작성을 방지했습니다!!

hyunw9 replied:
확인했습니다 !

hyunw9 commented on infra/src/main/java/sopt/makers/authentication/infra/client/openfeign/config/ClientProperty.java:
```diff
@@ -0,0 +1,14 @@
+
+@ConfigurationProperties("external.sopt-makers")
+public record ClientProperty(
+    Playground playground
```
이 부분은 PlaygroundProperty를 Record 타입으로 주입받아서 사용하는 부분일까요 ??

hyunw9 replied:
맞군요 ! 그렇다면 해당 프로퍼티는 플그쪽에 토큰값(인증여부) 반환을 위한 프로퍼티일까요 ?

yummygyudon replied:
기존 operation 프로젝트에서 사용하던 `ValueConfig.class`와 동일한 설정값 객체에용!!!
참고로 `@ConfigurationProperties` 어노테이션이 설정 파일을 읽어와 값을 Binding해주는 친구입니다.

그래서 아래 두 개의 객체는 동일하게 값을 가지게 된다고 생각하시면 됩니다!!
이렇게 적용한 이유는 PR 본문에 적어두었구요!!
```java
public class ValueConfig {
    
    @Value("${external.sopt-makers.playground.url}")
    private String playgroundUrl;
    ...

}

@ConfigurationProperties("external.sopt-makers")
public record ClientProperty(
    Playground playground
){
    public record Playground(
        String url
    ) {
    }
}
public 
```

hyunw9 reviewed (commented):
코드 한번 훑으면서 궁금한점 먼저 우선적으로 작성해보았습니다.. ! 답변 달아주시면 감사하겠습니다..! 

이건 개인적인 제안인데, 상수타입을 Const -> Constant로 변경하는건 어떨지 제안드립니다 ! 

정말 사소할 수 있는데, 개인적으로 더 잘 읽히는 느낌이긴 한 것 같아서요.. (Req, Res -> Response 의 느낌..? )

yummygyudon commented:
> 상수타입을 Const -> Constant로 변경하는건 어떨지 제안드립니다 !

동의합니다!! 리뷰 달아주시는대로 바로 반영하여 Commit 하겠습니다!

sung-silver commented on infra/src/main/java/sopt/makers/authentication/infra/client/openfeign/config/BearerAuthRequestInterceptor.java:
```diff
@@ -0,0 +1,17 @@
+
+    private final String token;
+
+    public BearerAuthRequestInterceptor(String token) {
```
P1: 생성자 파라미터에 final 키워드 반영 부탁드립니다!

sung-silver commented on infra/src/main/java/sopt/makers/authentication/infra/client/openfeign/config/BearerAuthRequestInterceptor.java:
```diff
@@ -0,0 +1,17 @@
+        this.token = token;
+    }
+
+    public abstract void apply(RequestTemplate template);
```
P1: 메서드 파라미터에 final 키워드 반영 부탁드립니다!

sung-silver commented on support/src/main/java/sopt/makers/authentication/support/system/security/filter/JwtAuthenticationFilter.java:
```diff
@@ -0,0 +1,44 @@
+    }
+
+    private String getAuthorizationToken(HttpServletRequest request) {
+        return request.getHeader(HttpHeaders.AUTHORIZATION).substring(HttpHeaders.AUTHORIZATION.length()).trim();
```
P5: return하는 값에 대해서 설명적 변수로 한번 명시하고 return해주면 리팩터링 과정에서 trim 메서드를 지운다던가 등의 실수를 줄일 수 있을 것 같은데 어떻게 생각하시나요?

yummygyudon replied:
한 줄로 처리하지 않고 `getHeader().substring()` 과 `trim()`을 분리하자는 말씀이시죠?!!
좋습니다!!

sung-silver replied:
```
String authorizationTokenFromHeader = request.getHeader(HttpHeaders.AUTHORIZATION).substring(HttpHeaders.AUTHORIZATION.length()));
String authorizationToken = authorizationTokenFromHeader.trim()
return authorizationToken
```

이런식으로 설명적 변수로 분리하는 것을 말씀드린겁니다!
trim을 분리하자는 말씀이신 것 같아서 분리해서 적어두긴했는데 한줄로 처리하고 trim까지 진행해야 토큰을 얻는다~는 느낌으로 가도 될 것 같네요!

sung-silver reviewed (changes requested):
저번주 final 키워드 붙이기로 한 내용이 전반적으로 반영되어있지 않은 것 같아서 확인 부탁드립니다!
저번주에 논의했던 내용 중 대부분을 구현해주셔서 너무 감사합니다 🥹

hyunw9 commented on support/src/main/java/sopt/makers/authentication/support/value/SystemConst.java:
```diff
@@ -0,0 +1,19 @@
+package sopt.makers.authentication.support.value;
+
+public final class SystemConst {
```
Const -> Constant 수정 요청입니다 !

hyunw9 commented on infra/src/main/java/sopt/makers/authentication/infra/value/ClientConst.java:
```diff
@@ -0,0 +1,8 @@
+package sopt.makers.authentication.infra.value;
+
+public final class ClientConst {
```
Const -> Constant 수정 요청입니다 !

hyunw9 reviewed (changes requested):
Const -> Constant 수정 요청

yummygyudon committed:
[STYLE] 변수명 & 패키지명 구체적 명시 및 변경

yummygyudon committed:
[REFACTOR] 메서드 파라미터 final 제어자 추가 및 불변성 보장

- Request 혹은 RequestTemplate 등의 객체와 같이 메서드 내에서 상태가 변경되는 객체의 경우, final을 부착하지 않았습니다.

yummygyudon committed:
[REFACTOR] 메서드 파라미터 final 제어자 추가 및 불변성 보장 (2)

sung-silver reviewed (approved):
진짜 고생하셨습니다~

yummygyudon committed:
[STYLE] token 추출 프로세스 명시를 위한 설명적 변수 반환으로 변경
